language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java

```diff
@@ -156,14 +155,13 @@ void setClientCapabilities(LSClientCapabilities clientCapabilities) {
 
     @Override
     public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(CompletionParams position) {
-        final List<CompletionItem> completions = new ArrayList<>();
         return CompletableFuture.supplyAsync(() -> {
             String fileUri = position.getTextDocument().getUri();
             Optional<Path> completionPath = CommonUtil.getPathFromURI(fileUri);
 
             // Note: If the source is a cached stdlib source or path does not exist, then return early and ignore
             if (completionPath.isEmpty() || CommonUtil.isCachedExternalSource(fileUri)) {
-                return Either.forLeft(completions);
+                return Either.forLeft(new ArrayList<>());
```

Comment:
Instead of returning `new ArrayList<>()` for erroneous cases we can return an unmodifiable per-declared empty list with `Collections..emptyList()`. This will limit ad-hoc new list objects creations.

Reply:
+1